tests/test-highlighting.zsh

```diff
@@ -106,62 +106,56 @@ run_test_internal() {
   # Check the data declares $PREBUFFER or $BUFFER.
   [[ -z $PREBUFFER && -z $BUFFER ]] && { echo >&2 "Bail out! On ${(qq)1}: Either 'PREBUFFER' or 'BUFFER' must be declared and non-blank"; return 1; }
   # Check the data declares $expected_region_highlight.
-  (( ${#expected_region_highlight} == 0 )) && { echo >&2 "Bail out! On ${(qq)1}: 'expected_region_highlight' is not declared or empty."; return 1; }
+  (( $+expected_region_highlight == 0 )) && { echo >&2 "Bail out! On ${(qq)1}: 'expected_region_highlight' is not declared."; return 1; }
 
   # Set sane defaults for ZLE variables
   : ${CURSOR=$#BUFFER} ${PENDING=0} ${WIDGET=z-sy-h-test-harness-test-widget}
 
   # Process the data.
   _zsh_highlight
 
-  # Overlapping regions can be declared in region_highlight, so we first build an array of the
-  # observed highlighting.
-  local i j
-  local -A observed_result
-  for ((i=1; i<=${#region_highlight}; i++)); do
-    local -a highlight_zone; highlight_zone=( ${(z)region_highlight[$i]} )
-    integer start=$highlight_zone[1] end=$highlight_zone[2]
-    if (( start < end )) # region_highlight ranges are half-open
-    then
-      (( --end )) # convert to closed range, like expected_region_highlight
-      (( ++start, ++end )) # region_highlight is 0-indexed; expected_region_highlight is 1-indexed
-      for j in {$start..$end}; do
-        observed_result[$j]=$highlight_zone[3]
-      done
-    else
-      # noop range; ignore.
-    fi
-    unset start end
-    unset highlight_zone
-  done
+  if (( unsorted )); then
+    region_highlight=("${(@n)region_highlight}")
+    expected_region_highlight=("${(@n)expected_region_highlight}")
```

Comment:
Suppose the input contains both `42 foo` and `42 bar`, is numeric sorting guaranteed to always output them in the same order, regardless of what other elements the input contains?

Plain old lexicographic sorting would definitely have this property but I suspect it would make for less nice failure modes (I presume we'd want failure to be reported on the "first" non-match, for some notion of "match" that depends on the start,stop indices).

Reply:
```
% a=(1 3 '2 foo' '2bar' '2 2 bar' '2 14 bar' '21 foo'); printf '<%s>' ${(@n)a} 
<1><2 2 bar><2 14 bar><2 foo><2bar><3><21 foo>
```

So it looks like numeric sorting falls back to lexicographic sorting.

Reply:
The implementation of the `(n)` sorting flag (NUMERIC_GLOB_SORT option) is:

https://github.com/zsh-users/zsh/blob/567733906210feaebce94675985eb1e77602bfd8/Src/sort.c#L119-L143

Reply:
In English:
1) Skip characters that are equal.
2) Store the comparison between the two different chars (This is how (n) falls back to lexicographical).
3) Assuming one of the differing characters is a digit, rewind to the start of digits.
4) Skip `0`s.
5) Skip equal digits.
6) Store the difference.
7) Find length and return in favor of the longer or the stored difference if equal.

So I think `(n)` DTRT.